feat: instrument multi-copy upload outcomes (#501)
filecoin-project/filecoin-pin: documentation/events-and-metrics.md

| <a id="tag-network"></a>`network` | `mainnet`, `calibration`, `devnet` | The Filecoin network the upload targeted. |

### Status Count Related Metrics

- These count metrics track the occurrence of a particular outcome for an event.
- All status count metrics carry an additional `value` tag that attributes the count to a specific outcome.

| Metric | Relevant Events | When Emitted | `value` Values | Source of truth |
|---|---|---|---|---|
| <a id="uploadCopyStatus"></a>`uploadCopyStatus` | [`uploadCopyResolved`](#uploadCopyResolved) | Once per copy in the upload result — one per entry in `result.copies` and one per entry in `result.failedAttempts`. | `success` (copy committed by the SP), `failure.pull` (Synapse pull step failed), `failure.commit` (Synapse commit step failed), `failure.other` (any other Synapse-reported failure). | [`src/core/telemetry/index.ts`](../src/core/telemetry/index.ts) |

In addition to the [common tags](#common-tags) and `value`, `uploadCopyStatus` carries:

- `spId` — [Storage Provider](glossary.md#service-provider) ID, stringified from the SDK's `providerId`.

> - `role` — `primary` or `secondary`, per the Synapse SDK.

Reviewer: should we have an `other` role for 3+ copies?

Author: Synapse only defines primary and secondary https://github.com/FilOzone/synapse-sdk/blob/e699e604f2c765100144efeed326181db17ee706/packages/synapse-sdk/src/storage/manager.ts#L93-L94

Reviewer: right. so do we flag n2+ all as "secondary"?

Author: It's ultimately out of scope for this PR, I think it only makes sense to follow Synapse here. To confirm: Any copy that is pulled from the primary will be called secondary, even if there are multiple secondary copies.

> To compute the **multi-copy success rate** over a window, divide `sum(uploadCopyStatus{value="success"})` by `sum(uploadCopyStatus)`. To isolate a particular failure mode (e.g. commit-step regressions across SPs), filter on `value="failure.commit"` and break down by `spId`.

Reviewer: FYI, I confirmed based on the existing data in betterstack that this will not work as documented.

```suggestion
To compute the **multi-copy success rate** over a window, divide the number of points tagged `value=success` by the total number of `uploadCopyStatus` points. To isolate a failure mode (e.g. commit-step regressions across SPs), filter on `value=failure.commit` and group by the `spId` tag.

> Querying in BetterStack: `uploadCopyStatus` is ingested as a counter, but each point is a flat `value:1`, so the default counter aggregation (`avgMerge(rate_avg)`) reads ~0. Count points with `sum(metrics_count)` and read tags via `label('value')` / `label('spId')` instead.
```

  participant Caller as Filecoin Pin caller<br/>(CLI / GitHub Action /<br/>Library / Website)
  participant Synapse as Synapse SDK
  participant SP as PDP Storage Provider

  Caller->>Synapse: uploadStarted
  loop For each copy
    Synapse->>SP: uploadCopyAttempted
    SP-->>Synapse: uploadCopyResolved (success or failure.*)
  end
  Synapse-->>Caller: uploadResultReceived
```

### Event List

| Event | Definition | Source of truth |
|------|------------|-----------------|
| <a id="uploadStarted"></a>`uploadStarted` | Filecoin Pin invokes `executeUpload` for one piece. | [`src/core/upload/index.ts`](../src/core/upload/index.ts) |
| <a id="uploadCopyAttempted"></a>`uploadCopyAttempted` | The Synapse SDK begins one copy attempt against a specific SP. Implicit marker — emitted once per entry that later shows up in `result.copies` or `result.failedAttempts`. | [`@filoz/synapse-sdk`](https://github.com/FilOzone/synapse-sdk) |
| <a id="uploadCopyResolved"></a>`uploadCopyResolved` | A copy attempt produces a terminal outcome (`success` or one of the `failure.*` values). Drives [`uploadCopyStatus`](#uploadCopyStatus). | [`src/core/telemetry/index.ts`](../src/core/telemetry/index.ts) |

> | <a id="uploadResultReceived"></a>`uploadResultReceived` | `executeUpload` returns with `{copies, failedAttempts}`. This is the point at which `recordUploadResult` is called and the [`uploadCopyStatus`](#uploadCopyStatus) points are submitted. | [`src/core/upload/index.ts`](../src/core/upload/index.ts) |

Reviewer 2: > recordUploadResult

This is the only place this is mentioned. From this doc, it wasn't obvious what this is.

Reviewer 2: > uploadCopyStatus points are submitted

I don't know what these "`uploadCopyStatus` points" are.

Author: > > recordUploadResult
> 
> This is the only place this is mentioned. From this doc, it wasn't obvious what this is.

That's the main function introduced in this PR, submitting upload results. It's an internal function.

> > uploadCopyStatus points are submitted
> 
> I don't know what these "`uploadCopyStatus` points" are.

See `### Status Count Related Metrics`

Reviewer 2: ```suggestion
| <a id="uploadResultReceived"></a>`uploadResultReceived` | `executeUpload` returns with `{copies, failedAttempts}`. This is the point at which `recordUploadResult` is called and the [`uploadCopyStatus`](#uploadCopyStatus) data points are submitted. | [`src/core/upload/index.ts`](../src/core/upload/index.ts) |
```

I think I had confusion with the 
>"uploadCopyStatus" points
(particularly the "points" portion)

I was confused because of 
1. English "point in time"
2. BetterStack doesn't use the term in https://betterstack.com/docs/logs/ingesting-data/http/metrics/

I think we should use "metric" (or ingested metric) in events-and-metrics.md or use "data point" if want want OTel familiarity.

For now I just suggested "data point".

# Filecoin Pin Events & Metrics

This document is the intended **source of truth** for the events emitted by Filecoin Pin and the metrics computed from them. It is intended for Filecoin Pin dashboard consumers and maintainers who need to understand what each metric means and where it comes from.

> **Note on "events":** the entries in the [Event List](#event-list) are named **timing markers** used to define when metrics are emitted — they are not all emitted as discrete events or log lines. Each marker is anchored in code (as a timestamp variable, log line, or status transition) and used to compute the metrics in the [Metrics](#metrics) section.

## Upload Event Model

> An upload (`executeUpload` in the [JavaScript library](../README.md#-javascript-library)) asks the [Synapse SDK](glossary.md#synapse) to replicate the same [Piece](glossary.md#piece) across one primary and zero or more secondary copies. Each copy is an independent attempt against a single [Storage Provider](glossary.md#service-provider); the SDK reports one outcome per copy in a single result object. Filecoin Pin records one telemetry event per resolved copy outcome.

Reviewer 2: I realize now we're being consistent with Synapse terminology, but the word "copy" was confusing me when thinking about these metrics.  As a result, the text below is an attempt to clarify it.

```suggestion
An upload (`executeUpload` in the [JavaScript library](../README.md#-javascript-library)) asks the [Synapse SDK](glossary.md#synapse) to replicate the same [Piece](glossary.md#piece) from Synapse across one primary and zero or more secondary copies. A "copy" is an independent attempt against a single [Storage Provider](glossary.md#service-provider). Even the the initial upload from the Synapse client to a Storage Provider is a "copy". The SDK reports one outcome per "copy" in a [single result object](https://docs.filecoin.cloud/reference/filoz/synapse-sdk/synapse/interfaces/uploadresult/#copies). Filecoin Pin records one telemetry event per resolved "copy" outcome. (Learn more about [copies in FOC](glossary.md#piece-copy).)
```

I'm thinking we could also add something like the following to the glossary:

```
## Piece Copy
In Filecoin Pin and Synapse, a copy is one independent storage placement of the same Piece on a single Service Provider (usually in its own Data Set). All copies share the same PieceCid; they differ by which SP holds the bytes and records the piece on chain.

This is not the everyday sense of “original plus one duplicate.” Here, N copies means N stored instances across N providers, not N extras on top of an unnamed original. Even the first upload target is called a copy with role "primary".  (See [Synapse upload docs](https://docs.filecoin.cloud/reference/filoz/synapse-sdk/storage/classes/storagemanager/#upload).)
```

Author: Done in e6b545e — applied your suggested rewording to the Upload Event Model paragraph (with the link to the new glossary entry) and added the `Piece Copy` glossary entry. Thanks!

### Upload Event Timeline

```mermaid
sequenceDiagram
  autonumber
  participant Caller as Filecoin Pin caller<br/>(CLI / GitHub Action /<br/>Library / Website)
  participant Synapse as Synapse SDK
  participant SP as PDP Storage Provider

  Caller->>Synapse: uploadStarted

>   loop For each copy

Reviewer 2: ```suggestion
  loop For each copy (primary or secondary)
```

  end
  Synapse-->>Caller: uploadResultReceived
```

### Event List

| Event | Definition | Source of truth |
|------|------------|-----------------|
| <a id="uploadStarted"></a>`uploadStarted` | Filecoin Pin invokes `executeUpload` for one piece. | [`src/core/upload/index.ts`](../src/core/upload/index.ts) |
| <a id="uploadCopyAttempted"></a>`uploadCopyAttempted` | The Synapse SDK begins one copy attempt against a specific SP. Implicit marker — emitted once per entry that later shows up in `result.copies` or `result.failedAttempts`. | [`@filoz/synapse-sdk`](https://github.com/FilOzone/synapse-sdk) |
| <a id="uploadCopyResolved"></a>`uploadCopyResolved` | A copy attempt produces a terminal outcome (`success` or one of the `failure.*` values). Drives [`uploadCopyStatus`](#uploadCopyStatus). | [`src/core/telemetry/index.ts`](../src/core/telemetry/index.ts) |
| <a id="uploadResultReceived"></a>`uploadResultReceived` | `executeUpload` returns with `{copies, failedAttempts}`. This is the point at which `recordUploadResult` is called and the [`uploadCopyStatus`](#uploadCopyStatus) points are submitted. | [`src/core/upload/index.ts`](../src/core/upload/index.ts) |

## Metrics

> Metrics ship as [direct HTTP POSTs](https://betterstack.com/docs/logs/ingesting-data/http/metrics/) to BetterStack from [`src/core/telemetry/index.ts`](../src/core/telemetry/index.ts).

Reviewer 2: ```suggestion
Metrics ship as [direct HTTP POSTs](https://betterstack.com/docs/logs/ingesting-data/http/metrics/) to BetterStack from [`src/core/telemetry/index.ts`](../src/core/telemetry/index.ts). One "upload" typically produces several metric data points (one per "copy" outcome).
```

### Common Tags

Every metric emitted by Filecoin Pin carries the following tags:

| Tag | Values | Definition |
|---|---|---|

> | <a id="tag-affordance"></a>`affordance` | `CLI`, `GitHub Action`, `Library`, `Filecoin Pin Website` | Which [affordance](../README.md#affordances) emitted the metric. Set by the host via `configureTelemetry({ affordance })`; defaults to `Library` when not configured. `configureTelemetry` rejects any other value at runtime. |

Reviewer 2: ```suggestion
| <a id="tag-affordance"></a>`affordance` | `CLI`, `GitHub Action`, `Library`, `Website` | Which [affordance](../README.md#affordances) emitted the metric. Set by the host via `configureTelemetry({ affordance })`; defaults to `Library` when not configured. `configureTelemetry` rejects any other value at runtime. |
```

For CLI, GitHub Action, and Library, the "Filecoin Pin" portion is implied. Do we do the same for the website? Or maybe for the website we do pin.filecoiin.cloud to be clear?

What's the reason to reject any other values? What do we gain with the strictness?

Reviewer 2: Also, will there be a PR in filecoin-pin-website after this?

Author: +1 renamed affordance to `pin.filecoin.cloud` in d3fb5a3 e365b07

> What's the reason to reject any other values? What do we gain with the strictness?

It's to ensure that one doesn't submit with a typo, mixed case, etc.

> Also, will there be a PR in filecoin-pin-website after this?

Good catch! https://github.com/filecoin-project/filecoin-pin-website/issues/187

| <a id="tag-network"></a>`network` | `mainnet`, `calibration`, `devnet` | The Filecoin network the upload targeted. |

### Status Count Related Metrics

- These count metrics track the occurrence of a particular outcome for an event.
- All status count metrics carry an additional `value` tag that attributes the count to a specific outcome.

| Metric | Relevant Events | When Emitted | `value` Values | Source of truth |
|---|---|---|---|---|

> | <a id="uploadCopyStatus"></a>`uploadCopyStatus` | [`uploadCopyResolved`](#uploadCopyResolved) | Once per copy in the upload result — one per entry in `result.copies` and one per entry in `result.failedAttempts`. | `success` (copy committed by the SP), `failure.pull` (Synapse pull step failed), `failure.commit` (Synapse commit step failed), `failure.other` (any other Synapse-reported failure). | [`src/core/telemetry/index.ts`](../src/core/telemetry/index.ts) |

Reviewer 2: ```suggestion
| <a id="uploadCopyStatus"></a>`uploadCopyStatus` | [`uploadCopyResolved`](#uploadCopyResolved) | Once per copy in the upload result — one per entry in `result.copies` and one per entry in `result.failedAttempts`. | `success` (copy committed by the SP), `failure.commit` (Synapse commit step failed),`failure.pull` (Synapse pull step failed), `failure.other` (any other Synapse-reported failure). | [`src/core/telemetry/index.ts`](../src/core/telemetry/index.ts) |
```

(nit) but putting commit before pull because the pull usually happens after commit.

Reviewer 2: I'm good to ignore this. I was thinking about the "primary" case as the first case where there isn't "pull". But for the "secondary" case, your order is right. I"m fine to ignore this.

In addition to the [common tags](#common-tags) and `value`, `uploadCopyStatus` carries:

- `spId` — [Storage Provider](glossary.md#service-provider) ID, stringified from the SDK's `providerId`.
- `role` — `primary` or `secondary`, per the Synapse SDK.

> To compute the **multi-copy success rate** over a window, divide the number of points tagged `value=success` by the total number of `uploadCopyStatus` points. To isolate a failure mode (e.g. commit-step regressions across SPs), filter on `value=failure.commit` and group by the `spId` tag.

Reviewer 2: I don't love this from the regard that it isn't answering: "what proportion of golden-path 'adds' succeed" where "succeed" means all the copies are stored. 

If someone specifies only one copy, that shouldn't pollute the "golden path" metric. Similarly, if someone specifies they want 100 copies but 99 of those fail, I don't want that tanking the overall success rate of the multi-copy flow. 

This was the motivation behind the `multicopy` proposal in https://github.com/filecoin-project/filecoin-pin/issues/363#issuecomment-4514202570

Does that make sense? 

Anyways, from a pragmatic regard though, what you have is probably good enough for now. I think I would then just want to document it accordingly.

```suggestion
To compute the **upload success rate of all copies (independent of "primary" or "secondary")** over a window, divide the number of data points tagged `value=success` by the total number of `uploadCopyStatus` data points. To isolate a failure mode (e.g. commit-step regressions across SPs), filter on `value=failure.commit` and group by the `spId` tag.

Note that this computation doesn't directly answer "what proportion of golden-path 'adds' succeed" since:
1. "adds" that only want a single upload (no extra copies) will be included AND
2. "adds" that specify more than a single extra copy will add more weight to the metric
[#516](https://github.com/filecoin-project/filecoin-pin/issues/516) is tracking more accurately answer this question.
```

Author: Agreed, good enough for now, and straight forward to improve later

> > Querying in BetterStack: `uploadCopyStatus` is ingested as a counter, but each point is a flat `value:1`, so the default counter aggregation (`avgMerge(rate_avg)`) reads ~0. Count points with `sum(metrics_count)` and read tags via `label('value')` / `label('spId')` instead.

Reviewer 2: This is probably my ignorance, but is the `value:1` different than the `value` label we have defined?

I'm also not following why something that is a flat `value:1` would have an average that read as ~0 instead of ~1.

------
Oh from looking at https://betterstack.com/docs/logs/ingesting-data/http/metrics/ I see the "value" is a first class term for the counter. Maybe we shouldn't do this to avoid confusion between counter "value" and label/tag "value"? 

I think this goes away if we go with my idea below of just having a single `uploadCopyBytes` gauge metric. 

Or if we don't like that, maybe we don't use a label/tag called "value" but instead set the label/tag as "status"? That deviates from what dealbot does though...

Reviewer 2: I keep feeling torn on what to do here. I think it's most clear if we change from using the label/tag of "value" to instead be "status".

Author: ba271d6

### Gauge Metrics

- These metrics carry a single numeric reading per point.

> - Each gauge point is emitted alongside the corresponding [`uploadCopyStatus`](#uploadCopyStatus) point, sharing its tag set so the two metrics can be joined at query time.

Reviewer 2: I'm wondering if instead of `uploadCopyStatus` and `uploadCopySize` we just do `uploadCopyBytes` (rename of `uploadCopyBytes`). I think `uploadCopyBytes` alone can give us everything we're wanting here. We then have a label/tag for "status" which can be `success` (copy committed by the SP), `failure.commit` (Synapse commit step failed),`failure.pull` (Synapse pull step failed), `failure.other` (any other Synapse-reported failure).

Does that make sense?

Reviewer 2: Stepping away and coming back to this with a fresh head, I think this is probably a bad idea.

A gauge alone makes counting successes awkward; the counter+gauge pair is dealbot-aligned and supports “count failures” and “p99 failure size” cleanly. If anything, maybe we just rename the "value" label/tag to "status" as the highest-leverage fix.

Author: ba271d6

| Metric | Relevant Events | When Emitted | Gauge Value | Source of truth |
|---|---|---|---|---|

> | <a id="uploadCopySize"></a>`uploadCopySize` | [`uploadCopyResolved`](#uploadCopyResolved) | Once per copy in the upload result, paired with [`uploadCopyStatus`](#uploadCopyStatus). | Piece size in bytes (`UploadResult.size`). All copies of one upload share the same size — the value identifies the upload that produced the outcome. | [`src/core/telemetry/index.ts`](../src/core/telemetry/index.ts) |

Reviewer 2: ```suggestion
| <a id="uploadCopyBytes"></a>`uploadCopyBytes` | [`uploadCopyResolved`](#uploadCopyResolved) | Once per copy in the upload result, paired with [`uploadCopyStatus`](#uploadCopyStatus). | Piece size in bytes (`UploadResult.size`). All copies of one upload share the same size — the value identifies the upload that produced the outcome. | [`src/core/telemetry/index.ts`](../src/core/telemetry/index.ts) |
```

Make this self-documenting when looking at the metric name?

Author: Done in 10b3390 — renamed `uploadCopySize` → `uploadCopyBytes` so the unit is self-documenting in the metric name (Prometheus/OTel `*Bytes` convention), updated the constant, emitter, tests, and this doc. Thanks!
